cmd/kubectl_token.go

```diff
@@ -0,0 +1,594 @@
+	}
+	defer f.Close()
+	return json.NewEncoder(f).Encode(cred)
+}
+
+func loginAndGenerateCred(input *LoginInput) (*ExecCredential, error) {
+	if input.authProvider == "" {
+		provider, err := getAuthProvider(input.server)
+		if err != nil {
+			return nil, err
+		}
+		input.authProvider = provider
+	}
+	tlsConfig, err := getTLSConfig(input)
+	if err != nil {
+		return nil, err
+	}
+	token := managementClient.Token{}
+	if samlProviders[input.authProvider] {
+		token, err = samlAuth(input, tlsConfig)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		customPrint(fmt.Sprintf("Enter credentials for %s \n", input.authProvider))
+		token, err = promptAuth(input, tlsConfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cred := &ExecCredential{
+		TypeMeta: TypeMeta{
+			Kind:       "ExecCredential",
+			APIVersion: "client.authentication.k8s.io/v1beta1",
+		},
+		Status: &ExecCredentialStatus{},
+	}
+	cred.Status.Token = token.Token
+	if token.ExpiresAt == "" {
+		return cred, nil
+	}
+	ts, err := time.Parse(time.RFC3339, token.ExpiresAt)
+	if err != nil {
+		customPrint(fmt.Sprintf("\n error parsing time %s %v", token.ExpiresAt, err))
+		return nil, err
+	}
+	cred.Status.ExpirationTimestamp = &Time{Time: ts}
+	return cred, nil
+
+}
+
+func promptAuth(input *LoginInput, tlsConfig *tls.Config) (managementClient.Token, error) {
+	token := managementClient.Token{}
+	username, err := customPrompt("username", true)
+	if err != nil {
+		return token, err
+	}
+
+	password, err := customPrompt("password", false)
+	if err != nil {
+		return token, err
+	}
+
+	responseType := "kubeconfig"
+	if input.clusterID != "" {
+		responseType = fmt.Sprintf("%s_%s", responseType, input.clusterID)
+	}
+
+	body := fmt.Sprintf(`{"responseType":"%s", "username":"%s", "password":"%s"}`, responseType, username, password)
+
+	url := fmt.Sprintf("%sv3-public/%ss/%s?action=login", input.server, input.authProvider,
+		strings.ToLower(strings.Replace(input.authProvider, "Provider", "", 1)))
+
+	response, err := request(http.MethodPost, url, bytes.NewBufferString(body))
+	if err != nil {
+		return token, nil
+	}
+
+	err = json.Unmarshal(response, &token)
+	if err != nil {
+		return token, err
+	}
+	return token, nil
+}
+
+func samlAuth(input *LoginInput, tlsConfig *tls.Config) (managementClient.Token, error) {
+	token := managementClient.Token{}
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return token, err
+	}
+	publicKey := privateKey.PublicKey
+	marshalKey, err := json.Marshal(publicKey)
+	if err != nil {
+		return token, err
+	}
+	encodedKey := base64.StdEncoding.EncodeToString(marshalKey)
+
+	id, err := generateKey()
+	if err != nil {
+		return token, err
+	}
+
+	responseType := "kubeconfig"
+	if input.clusterID != "" {
+		responseType = fmt.Sprintf("%s_%s", responseType, input.clusterID)
+	}
+
+	tokenURL := fmt.Sprintf(authTokenURL, input.server, id)
+
+	req, err := http.NewRequest(http.MethodGet, tokenURL, bytes.NewBuffer(nil))
```

Comment:
this is supposed to be get and not post ?

Reply:
yep, this is polling `v3-public/authtokens` to read the encrypted token after successful saml login